jsonrpc/http.go

```diff
@@ -30,31 +31,51 @@ func NewHTTP(rpc *Server, log utils.SimpleLogger) *HTTP {
 	return h
 }
 
-// ServeHTTP processes an incoming HTTP request
-func (h *HTTP) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
+type httpConn struct {
+	reqBody io.ReadCloser
+	rw      http.ResponseWriter
+}
+
+func accept(rw http.ResponseWriter, req *http.Request) *httpConn {
 	if req.Method == "GET" {
 		status := http.StatusNotFound
 		if req.URL.Path == "/" {
 			status = http.StatusOK
 		}
-		writer.WriteHeader(status)
-		return
+		rw.WriteHeader(status)
+		return nil
 	} else if req.Method != "POST" {
-		writer.WriteHeader(http.StatusMethodNotAllowed)
-		return
+		rw.WriteHeader(http.StatusMethodNotAllowed)
+		return nil
+	}
+
+	return &httpConn{
+		reqBody: http.MaxBytesReader(rw, req.Body, MaxRequestBodySize),
+		rw:      rw,
+	}
+}
+
+func (c *httpConn) Read(p []byte) (int, error) {
+	return c.reqBody.Read(p)
+}
+
+// Write returns the number of bytes of p written, not including the header.
+func (c *httpConn) Write(p []byte) (int, error) {
+	c.rw.Header().Set("Content-Type", "application/json")
+	if p == nil {
+		return 0, nil
 	}
+	return c.rw.Write(p)
+}
 
-	req.Body = http.MaxBytesReader(writer, req.Body, MaxRequestBodySize)
+// ServeHTTP processes an incoming HTTP request
+func (h *HTTP) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
+	conn := accept(writer, req)
+	if conn == nil {
+		return
+	}
 	h.requests.Inc()
-	resp, err := h.rpc.HandleReader(req.Context(), req.Body)
-	writer.Header().Set("Content-Type", "application/json")
-	if err != nil {
+	if err := h.rpc.Handle(req.Context(), conn); err != nil {
 		writer.WriteHeader(http.StatusInternalServerError)
```

Comment:
this potentially brings back the error fixed [here](https://github.com/NethermindEth/juno/pull/1075)
First call to `conn.Write` in `server.Handle` might successfully write a 200 header and fail to write the body. Then here we will try to write a 500 header again.